### PR TITLE
feat(stream): Make scale DAG aware

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

**This section will be used as the commit message. Please do not leave this empty!**

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Let `Reschedule` accept more than one downstream fragment.
- Let the dispatcher id calculated from the exchange operator id together with the upstream and downstream id.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
https://github.com/risingwavelabs/risingwave/issues/6955

### CODE REVIEW

File: src/meta/src/stream/stream_graph.rs [hunk 1]
```diff
@@ -362,10 +362,22 @@ impl StreamGraphBuilder {
...
+        let edge_upstream_id = edge.upstream_id;
+        let edge_downstream_id = edge.downstream_id;
+        // We can't use the exchange operator id directly as the dispatch id, because an exchange
+        // could belong to more than one downstream in DAG. We can use the exchange operator
+        // id together with edge upstream and downstream id as an unique id. In this way we
+        // can ensure the dispatchers of `StreamActor` would have different id, even though they
+        // come from the same exchange operator.
+        let dispatch_id = exchange_operator_id
+            + edge_upstream_id as u64 * 1000
+            + edge_downstream_id as u64 * 1000000;
```
Comment (BugenZhao) [inline src/meta/src/stream/stream_graph.rs:379]: IIRC, the dispatcher id has to be unique only in the scope of this actor. So can we directly use the downstream fragment id here?

https://github.com/risingwavelabs/risingwave/blob/181680b58fdbbadf49f8c672d3fc8d43fe84d6c0/proto/stream_plan.proto#L528-L531

For example, MV on MV uses downstream actor ID (actually fragment ID should be enough).

https://github.com/risingwavelabs/risingwave/blob/181680b58fdbbadf49f8c672d3fc8d43fe84d6c0/src/meta/src/stream/stream_manager.rs#L253-L255

File: src/meta/src/stream/scale.rs [hunk 2]
```diff
@@ -1102,7 +1100,7 @@ where
                     vnode_bitmap_updates,
                     upstream_fragment_dispatcher_ids,
                     upstream_dispatcher_mapping,
-                    downstream_fragment_id,
+                    downstream_fragment_ids: downstream_fragment_id,
```
Comment (BugenZhao) [inline src/meta/src/stream/scale.rs:1103]:
```suggestion
                    downstream_fragment_ids,
```

File: src/tests/simulation/src/nexmark.rs [hunk 3]
```diff
@@ -514,6 +514,27 @@ SELECT * FROM nexmark_q105;
...
+    pub mod q106 {
+        //! A self-made query that covers singleton top-n (and local-phase group top-n).
+        //!
+        //! Show the top 1000 auctions by the number of bids.
+
+        use super::*;
+        pub const CREATE: &str = r#"
+CREATE MATERIALIZED VIEW nexmark_q106
+AS
+select count(*) cnt from auction A join auction B on A.id = B.id;
```
Comment (BugenZhao) [inline src/tests/simulation/src/nexmark.rs:531]: What's this case for? 👀

File: src/meta/src/stream/stream_graph.rs (hunk 1 above)
Comment (chenzl25) [reply src/meta/src/stream/stream_graph.rs:379]: You are right, we can use edge downstream_id as dispatcher id directly.

File: src/tests/simulation/src/nexmark.rs (hunk 3 above)
Comment (chenzl25) [reply src/tests/simulation/src/nexmark.rs:531]: Just a simplified test case. I will remove it.

**Review by BugenZhao — APPROVED** [review]
LGTM. I didn't realize the changes could be so minor. 🥺

**Pull request merged**